### PR TITLE
Add extended buttons to support trackpad (#191)

### LINKED ISSUE #191: Add trackpad button support
The current W3C Gamepad specification does not include mappings for the trackpad button on Sony DualShock and DualSense controllers. This omission leads to inconsistencies across browsers, as observed in WebKit (Safari) and Chromium (Chrome).

### Suggested Update:
1. **Standardize Trackpad Button Mapping:** Introduce a standard mapping for the trackpad button, possibly as button 17, to align with Chromium's existing implementation.
2. **Uniform Button Array Size:** Ensure a 17-element button array for all gamepad controllers to maintain consistency and aid in fingerprinting mitigation.

### Additional Information:
This inconsistency was noted in WebKit bug report [266293](https://bugs.webkit.org/show_bug.cgi?id=266293), with a test conducted on gamepad-tester.com showing different behaviors in Safari Technology Preview and Chrome.

### Impact:
Addressing this will enhance cross-browser compatibility and user experience for gamepad users, especially those using Sony's controllers.

**References:**
- WebKit Bug Report: [266293 – Gamepad: Map DualShock/DualSense Trackpad Button](https://bugs.webkit.org/show_bug.cgi?id=266293)

### PR DESCRIPTION
**Pull request by xingri**
Closes #191 

The following tasks have been completed:

 * [ ] Modified Web platform tests (link to pull request)

Implementation commitment:

 * [ ] WebKit (https://bugs.webkit.org/show_bug.cgi?id=266293)
 * [ ] Chromium (https://issues.chromium.org/issues/339841686)
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)

***
<a href="https://pr-preview.s3.amazonaws.com/xingri/gamepad/pull/196.html" title="Last updated on Jul 10, 2025, 11:10 PM UTC (eeea475)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/gamepad/196/c430e73...xingri:eeea475.html" title="Last updated on Jul 10, 2025, 11:10 PM UTC (eeea475)">Diff</a>

### CODE REVIEW

File: extensions.html [hunk 1]
```diff
@@ -649,6 +649,168 @@ <h2>
         </p>
       </dd>
     </section>
+    <section>
```
Comment (nondebug) [inline extensions.html]: Let's define extended buttons in the main spec document (index.html). The extensions document is not officially part of the spec and we want to eventually remove it.

The spec PR should extend [GamepadButton](https://www.w3.org/TR/gamepad/#gamepadbutton-interface) to add a new attribute (perhaps called `GamepadButton.name`) and an algorithm for the getter. The type of the new attribute will be a new enum (perhaps called `GamepadButtonName`). The enum type should be defined in the spec along with all its values.

Add steps to initialize the new attribute when [initializing buttons](https://www.w3.org/TR/gamepad/#dfn-initializing-buttons). The steps should define what to do if a gamepad is missing a button defined in the Standard Gamepad. I think it makes sense to use a special value for these buttons so applications can identify which buttons are not actually mapped. (Ideally we'd set the button to null, but this would break existing apps which don't check for null.)

Also, initialization steps should define what to do if a gamepad has a button that doesn't have a predefined Standard Gamepad index or enum value. Do we enumerate it? If so, what should the new attribute's value be? Perhaps we can add a special "unrecognized" value for this.

File: extensions.html [hunk 2]
```diff
@@ -649,6 +649,168 @@ <h2>
...
+      <h2>
+        "extended" Gamepad Buttons Mapping
+      </h2>
+      <p>
+        This section defines extended gamepad buttons mapping beyonds the Standard Gamepads layout,
+        but are commonly found on some gamepad models .The <code>_ext</code> suffix is used to indicate
```
Comment (nondebug) [inline extensions.html]: The `_ext` suffix comes from OpenXR's input subpaths specification:

https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#_adding_input_sources_via_extensions

We should decide whether this attribute uses OpenXR input subpaths or something new that we want to define for ourselves. If we're using OpenXR subpaths then we should explicitly link to the OpenXR spec which explains how to extend it. If we're only using it for inspiration then I think we should drop the `_ext` suffix.

File: extensions.html [hunk 3]
```diff
@@ -649,6 +649,168 @@ <h2>
...
+            <td>dpad_right</td>
+            <td>Standard: Right button in left cluster</td>
+          </tr>
+          <tr>
+            <td>buttons[16]</td>
+            <td>system</td>
+            <td>Standard: Center button in center cluster</td>
+          </tr>
+          <tr>
+            <td>buttons[17]</td>
```
Comment (nondebug) [inline extensions.html]: This suggests the "trackpad" button can only appear at `buttons[17]`. My thinking is that we should not continue the current pattern of assigning meaning to button indices, otherwise the `buttons` array will be very sparse since most gamepads do not share the same extra buttons. Instead of using button indices to identify buttons, apps can check the new enum attribute.

For this non-normative section, I think it will be clearer if we provide a separate table for each gamepad instead of trying to combine them all into a single table.

Comment (xingri) [reply extensions.html]: @nondebug  Thanks for the review. Will update the PR based on the feedback shortly.

Comment (xingri) [reply extensions.html]: @nondebug I have updated the PR based on your feedback. I have added the enum for identifying the gamepad buttons. Please review this changes and share your opinions.

File: index.html [hunk 4]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+            This enum defines the Extended type of Gamepads.
+          </p>
+          <pre class="idl">
+            enum partial-GamepadMappingType {
+              "extended"
+            };
+          </pre>
+          <dl>
+            <dt>
+              <dfn>`"extended"`</dfn>
```
Comment (nondebug) [inline index.html]: Introducing a new value for GamepadMappingType will break applications that currently check for "standard". I don't think it's necessary to add a new enum value for this, applications can discover whether the gamepad supports extra buttons by inspecting the buttons array.

File: index.html [hunk 5]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+        <section>
+          <h2>
+            Partial <dfn>GamepadButton</dfn> Interface
+          </h2>
+          <p>
+            This partial interface extends the GamepadButton interface in this spec.
+          </p>
+          <pre class="idl">
+            partial interface GamepadButton {
+              readonly attribute GamepadButtonType type;
```
Comment (nondebug) [inline index.html]: Please move the definition to the main `interface GamepadButton` IDL section.

File: index.html [hunk 6]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+              <th>
+                Initial value
+              </th>
+              <th>
+                Description (non-normative)
+              </th>
+            </tr>
+            <tr>
+              <td>
+                <dfn data-dfn-for="GamepadButton">[[\type]]</dfn>
```
Comment (nondebug) [inline index.html]: I don't think we need an internal slot.

For other attributes we use a slot because the attribute is read-only but the implementation needs to be able to change the value. In this case the value should never change after initialization.

File: index.html [hunk 7]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+                `undefined`
+              </td>
+              <td>
+                Indicates the button type the controller is held in.
+              </td>
+            </tr>
+	  </table>
+          <dl data-dfn-for="GamepadButton">
+          <dt>
+            <dfn>type</dfn>
```
Comment (nondebug) [inline index.html]: `<dfn>type</dfn> attribute`

File: index.html [hunk 8]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+              The Gamepad's controls have been mapped to the Extended Gamepad layout.
+            </dd>
+          </dl>
+        </section>
+	<section>
+          <h3>
+            <dfn>GamepadButtonType</dfn> Enum
+          </h3>
+          <p>
+            This enum defines the set of possible botton types.
```
Comment (nondebug) [inline index.html]: botton -> button

File: index.html [hunk 9]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+          </p>
+          <pre class="idl">
+            enum GamepadButtonType {
+              "",  /* unknown, or not applicable */
+              "trackpad",
+              "share",
+              "..."
```
Comment (nondebug) [inline index.html]: Omit "..." since it is not a valid enum value.

File: sony_dualshock4_gamepad.svg
Comment (nondebug) [inline sony_dualshock4_gamepad.svg:1]: The diagrams are difficult to read. For clarity I think we should provide this information as a table instead of a diagram.

File: index.html [hunk 10]
```diff
@@ -1718,232 +1718,429 @@ <h2>
...
+                Indicates the button type the controller is held in.
+              </td>
+            </tr>
+	  </table>
+          <dl data-dfn-for="GamepadButton">
+          <dt>
+            <dfn>type</dfn>
+          </dt>
+          <dd>
+            An enumeration, {{GamepadButtonType}}, that indicates which button types the controller
```
Comment (nondebug) [inline index.html]: The description for the `type` attribute needs to be updated

Comment (xingri) [reply index.html]: @nondebug  Again thank you for the review. I have no objections and will update this PR based on your feedbacks.

File: index.html [hunk 11]
```diff
@@ -952,6 +953,17 @@ <h2>
...
         </tr>
+        <tr>
+          <td>
+            <dfn data-dfn-for="GamepadButton">[[\type]]</dfn>
+          </td>
+          <td>
+            `undefined`
+          </td>
+          <td>
+            Indicates the button type the controller is held in.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
            The button type of the controller.
```

File: index.html [hunk 12]
```diff
@@ -1718,232 +1737,341 @@ <h2>
...
+	<section>
+          <h3>
+            <dfn>GamepadButtonType</dfn> Enum
+          </h3>
+          <p>
+            This enum defines the set of possible button types.
+          </p>
+          <pre class="idl">
+            enum GamepadButtonType {
+              "",  /* unknown, or not applicable */
```
Comment (marcoscaceres) [inline index.html]: Better to make it explicit.

```suggestion
              "unknown",
```

And if not applicable, maybe we should make . type nullable?

File: index.html [hunk 13]
```diff
@@ -1718,232 +1737,341 @@ <h2>
...
-          </tr>
-        </tbody>
-      </table>
-      <figure>
-        <img src="standard_gamepad.svg" alt="">
-        <figcaption>
-          Visual representation of a [=Standard Gamepad=] layout.
-        </figcaption>
-      </figure>
+      <section>
```
Comment (marcoscaceres) [inline index.html:2470]: @xingri, would it be possible to revert some of this so it only shows what you changed?

Comment (xingri) [reply index.html:2470]: @marcoscaceres  thanks for the review. updated the changes by your feedback.

File: index.html [hunk 14]
```diff
@@ -1018,6 +1030,13 @@ <h2>
             </li>
           </ol>
         </dd>
+        <dt>
+          <dfn>type</dfn> attribute
+        </dt>
+        <dd>
+          An enumeration, {{GamepadButtonType}}, that indicates which button types the controller
+          is being held in.
+        </dd>
```
Comment (mrmcpowned) [inline index.html:1397]: Looks like we're still missing getter definitions, no?

File: index.html [hunk 15]
```diff
@@ -1018,6 +1030,13 @@ <h2>
             </li>
           </ol>
         </dd>
+        <dt>
+          <dfn>type</dfn> attribute
+        </dt>
+        <dd>
+          An enumeration, {{GamepadButtonType}}, that indicates which button types the controller
+          is being held in.
```
Comment (mrmcpowned) [inline index.html]: Is this phrasing in line with other specs? There's an [open spec issue](https://github.com/whatwg/html/issues/9832) regarding consistency in enum phrasing which makes me think this would be better phrased as:
> An enumerated {{GamepadButtonType}} attribute that classifies the current button's type in relation to an extended mapping

File: index.html [hunk 16]
```diff
@@ -1944,6 +1963,109 @@ <h2>
...
+          into the GamepadMappingType. Additionally, we have defined an enumeration for the various
+          button types in the table, termed "GamepadButtonType", and have expanded the "GamepadButton"
+          interface to encompass this new button type enumeration.
+       </p>
+        <table class="tg">
+          <thead>
+            <tr>
+              <th scope="col">Buttons</th>
+              <th scope="col">Gamepad Models</th>
+            </tr>
```
Comment (mrmcpowned) [inline index.html]: It feels like there's some disconnect with this table and the available GamepadButtonType enums. Are we missing a possible third column which maps each of these button to a respective GamepadButtonType?

File: index.html [hunk 17]
```diff
@@ -1944,6 +1963,109 @@ <h2>
...
+            <tr>
+              <td>squeeze_right_lower</td>
+              <td>Xbox Elite Wireless Controller</td>
+            </tr>
+          </tbody>
+        </table>
+       <section>
+          <h3>
+            <dfn>GamepadButtonType</dfn> Enum
```
Comment (mrmcpowned) [inline index.html:2485]: I see 8 different classes of extended buttons in the table above, but only 2 applicable GamepadButtonTypes in the enum definition. Is the intention to expand the enum list below so we have a 1:1 mapping with the rows in the table?

File: index.html [hunk 18]
```diff
@@ -1944,6 +1963,109 @@ <h2>
...
+            </tr>
+           <tr>
+              <td>voice assistant</td>
+              <td>Google Stadia Controller<br>
+                Amazon Luna Controller<br>
+              </td>
+            </tr>
+            <tr>
+              <td>home</td>
+              <td>Nvidia Shield Gamepad</td>
```
Comment (mrmcpowned) [inline index.html]: The Xbox and PlayStation controllers have their own respective home buttons. Are we intending to add those to this list or is there some special reason why we're classifying this separately for the Shield gamepad?

File: index.html [hunk 19]
```diff
@@ -1944,6 +1963,109 @@ <h2>
...
+              "trackpad",
+              "share"
+            };
+          </pre>
+          <dl data-dfn-for="GamepadButtonType">
+            <dt>
+              <dfn>unknown</dfn>
+            </dt>
+            <dd>
+              This indicates button type is unknown or not applicable.
```
Comment (mrmcpowned) [inline index.html]: Is there a tangible benefit for User Agents to populate all the GamepadButtons with `unknown` for any buttons that can't be classified as one of the other types? Would it be better to simply have the attribute as `undefined` when that's the case?

Comment (xingri) [reply index.html]: @mrmcpowned  Thanks for the review. I agree with your feedbacks. Will update this changes shortly.

File: index.html (hunk 15 above)
Comment (xingri) [reply index.html]: updated.

File: index.html (hunk 17 above)
Comment (xingri) [reply index.html:2485]: updated.

File: index.html (hunk 18 above)
Comment (xingri) [reply index.html]: Thanks. updated.

File: index.html (hunk 19 above)
Comment (xingri) [reply index.html]: Agree with your feedback @mrmcpowned, updated.

File: index.html [hunk 20]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+          <thead>
+            <tr>
+              <th scope="col">Buttons</th>
+              <th scope="col">Gamepad Models</th>
+            </tr>
+          </thead>
+          <tbody>
+            <tr>
+              <td>trackpad/touchpad</td>
+              <td>Sony DualSense, DualShock 4</td>
```
Comment (marcoscaceres) [inline index.html]: As I imagine these products will change a lot over time, so these should probably be in [Wiki page](https://github.com/w3c/gamepad/wiki) (or if we want to formalize it, and actual registry). We can point the spec to the wiki page.

File: index.html [hunk 21]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+              "squeeze_right",
+              "squeeze_right_lower"
+            };
+          </pre>
+          <dl data-dfn-for="GamepadButtonType">
+            <dt>
+              <dfn>trackpad</dfn>
+            </dt>
+            <dd>
+              This button is being assigned to trackpad button.
```
Comment (marcoscaceres) [inline index.html]: Probably just ok to say:
 
```suggestion
              A trackpad button.
```

Same for the rest.

File: index.html [hunk 22]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
       </figure>
+      <section>
+        <h3>
+          Extended Mapping
+        </h3>
+        <p>
+          This section introduces an extended gamepad button mapping beyond the Standard Gamepad layout.
+          These additional buttons are commonly found on certain gamepad models.
+          The following table defines the extended buttons used by some gamepad models in the following diagrams.
+          It’s important to note that this list is not exhaustive, and user agents may utilize different
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          It's important to note that this list is not exhaustive, and user agents may utilize different
```

File: index.html [hunk 23]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+          or additional buttons for these or other gamepad models.
+          Consequently, the number of buttons on the Gamepad is not limited to the standard mapping of 17 buttons.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          Consequently, the number of buttons on the {{Gamepad}} is not limited to the standard mapping of 17 buttons.
```

File: index.html [hunk 24]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+        </p>
+       <p>
+          To accommodate extended gamepads, we have incorporated an "extended" mapping enumeration
+          into the GamepadMappingType. Additionally, we have defined an enumeration for the various
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          into the {{GamepadMappingType}}. Additionally, we have defined an enumeration for the various
```

File: index.html [hunk 25]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+          button types in the table, termed "GamepadButtonType", and have expanded the "GamepadButton"
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          button types in the table, termed {{GamepadButtonType}}, and have expanded the {{GamepadButton}}
```

File: index.html [hunk 26]
```diff
@@ -1944,6 +1963,132 @@ <h2>
...
+       <section>
+          <h3>
+            <dfn>GamepadButtonType</dfn> Enum
+          </h3>
+          <p>
+            This enum defines the set of possible button types.
+          </p>
+          <pre class="idl">
+            enum GamepadButtonType {
+              "trackpad",
```
Comment (marcoscaceres) [inline index.html:2499]: Nit: Please alpha sort these

File: index.html (hunk 20 above)
Comment (xingri) [reply index.html]: @marcoscaceres  Thanks for the feedback. I have updated the rest of the comments.
I agree with you that it will keep changing and it would be great to have the link inside of the gamepad working group wiki. If this is not a hard requirement, I would like to follow up creating the wiki page on the different button mappings of the game controllers mentioned in this table as a separate PR.

File: index.html [hunk 27]
```diff
@@ -952,6 +953,17 @@ <h2>
             .. 1.0]
           </td>
         </tr>
+        <tr>
+          <td>
+            <dfn data-dfn-for="GamepadButton">[[\type]]</dfn>
```
Comment (nondebug) [inline index.html]: I think the internal slot [[type]] isn't needed and can be removed.

Normally we use internal slots for read-only attributes that the user agent needs to modify. We expect a GamepadButton created with a button type will always have that type, so we don't need a way to modify it.

File: index.html (hunk 20 above)
Comment (nondebug) [reply index.html]: sgtm, let's remove the table from this PR and follow up with the wiki page.

File: index.html
Comment (nondebug) [inline index.html:1]: Please add steps to the "initialize buttons" steps to initialize the button type.

https://w3c.github.io/gamepad/#dfn-initializing-buttons

Currently we initialize buttons in two phases. First, determine how many buttons are needed. Then, create all the buttons.

We'll need to remember the button type from the first phase so we can initialize the GamepadButton in the second phase. We also need to handle some special cases:

A GamepadButton may be created without a corresponding button input on the gamepad. (For instance, a gamepad may not have all the Standard Gamepad buttons; the button array entry for a missing button will still have a GamepadButton object but its value should never change.) We can initialize GamepadButton.type to null to represent this case.

A gamepad may have buttons that don't have a canonical button type. We should create a special GamepadButtonType enum value to signal that the button exists but doesn't have a canonical name yet.

Comment (xingri) [reply index.html:1]: @nondebug  Thanks for the feedback. Added steps for considering the button types.

File: index.html [hunk 28]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
         </figcaption>
       </figure>
+      <section data-dfn-for="ExtendedMapping" data-link-for="ExtendedMapping">
+        <h3>
+	  <dfn>ExtendedMapping</dfn>
+        </h3>
+        <p>
+          This section introduces an extended gamepad button mapping beyond the Standard Gamepad layout.
+          These additional buttons are commonly found on certain gamepad models.
+          The following table defines the extended buttons used by some gamepad models.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          The following definition list defines the <dfn>extended gamepad buttons</dfn> used by some gamepad models.
```

File: index.html [hunk 29]
```diff
@@ -820,8 +820,13 @@ <h3>
           <li>[=list/For each=] |rawInputIndex:long| of [=the range=] from 0 to
           |inputCount| − 1:
             <ol>
-              <li>If the the gamepad button at index |rawInputIndex|
-              [=represents a Standard Gamepad button=]:
+              <li>Determine the type of the button at index |rawInputIndex|
+              [=represents a Standard Gamepad button=] or Extended button
+              types as described in the {{ExtendedMapping}} using the provided
+              buttonTypes array. If the button does not have a canonical type,
+              use GamePadButtonType.non_canonical.
```
Comment (marcoscaceres) [inline index.html]: This should be specified as: 

```suggestion
              <li>Let |button| be the button at index |rawInputIndex|.</li>
              <li>Let |type| be the result of determining if the |button|'s type represents a [=Standard Gamepad button|represents a Standard Gamepad button=] or an [=Extended Gamepad button|represents a Extended Gamepad button=]. If it's neither, use {{GamePadButtonType/"non_canonical"}}. 
              </li>
```

And `<dfn>represents a Standard Gamepad button</dfn>` should be defined underneath where "represents a Standard Gamepad button" is defined/

File: index.html [hunk 30]
```diff
@@ -820,8 +820,13 @@ <h3>
...
+              </li>
+              <li>If the button type is not GamepadButtonType.non_canonical:
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
              <li>If |type| is not {{GamepadButtonType/"non_canonical"}}:
```

File: index.html [hunk 31]
```diff
@@ -878,8 +883,15 @@ <h3>
           <li>Initialize |buttons| to be an empty [=list=].
           </li>
           <li>[=list/For each=] |buttonIndex:long| of [=the range=] from 0 to
-          |buttonsSize| − 1, [=list/append=] a [=new=] {{GamepadButton}} to
-          |buttons|.
+          |buttonsSize| − 1:
+            <ol>
+              <li>Determine the type of the button from the provided buttonTypes
```
Comment (marcoscaceres) [inline index.html]: This should copy the steps above, with the correct linking.

File: index.html (hunk 14 above)
Comment (marcoscaceres) [reply index.html:1397]: Yeah, though it's implied... but yeah, if the others have the getter steps (what it was initialized to) then so should this.

File: index.html [hunk 32]
```diff
@@ -2031,6 +2051,62 @@ <h2>
           Visual representation of a [=Standard Gamepad=] layout.
         </figcaption>
       </figure>
+      <section data-dfn-for="ExtendedMapping" data-link-for="ExtendedMapping">
+        <h3>
+	  <dfn>ExtendedMapping</dfn>
+        </h3>
+        <p>
```
Comment (marcoscaceres) [inline index.html]: Oh wait. ExtendedMapping is not an IDL thing... it's a concept. 

```suggestion
      <section>
        <h3>
	  <dfn>Extended Gamepad Mapping</dfn>
        </h3>
        <p>
```

File: index.html [hunk 33]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+          This section introduces an extended gamepad button mapping beyond the Standard Gamepad layout.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          This section introduces an extended gamepad button mapping beyond the [=Standard Gamepad mapping=].
```

File: index.html [hunk 34]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+          These additional buttons are commonly found on certain gamepad models.
+          The following table defines the extended buttons used by some gamepad models.
+          It's important to note that this list is not exhaustive, and user agents may utilize different
+          or additional buttons for these or other gamepad models.
+          Consequently, the number of buttons on the {{Gamepad}} is not limited to the standard mapping of 17 buttons.
+        </p>
+       <p>
+          To accommodate extended gamepads, we have incorporated an "extended" mapping enumeration
```
Comment (marcoscaceres) [inline index.html]: It seems the "extended" value hasn't been added to GamepadMappingType yet. Can you please update the PR and add it,  along with this definition?

File: index.html [hunk 35]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+          </h3>
+          <p>
+            This enum defines the set of possible button types.
+          </p>
+          <pre class="idl">
+            enum GamepadButtonType {
+              "standard",
+              "non-canonical",
+               "trackpad", 
+               // ... More enum types to be defined by the separate W3C wiki page.
```
Comment (marcoscaceres) [inline index.html]: We can do this, but we need to change GamepadButtonType to an DOMString. Otherwise, the enum will always be out of date. Alternatively, we don't use the Wiki and just use the spec... it might be better to just use the spec, so we can have more oversight over what is defined here (and get better interop).

File: index.html [hunk 36]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+            };
+          </pre>
+          <dl data-dfn-for="GamepadButtonType">
+            <dt>
+              <dfn>standard</dfn>
+            </dt>
+            <dd>
+              Represent a button has a button type defined in the standard mapping.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
              Represent a button has a button type defined in the [=standard gamepad mapping=].
```

File: index.html [hunk 37]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+            </dd>
+            <dt>
+              <dfn>non-canonical</dfn>
+            </dt>
+            <dd>
+              Represents a button that exists but doesn't have a canonical name.
+            </dd>
```
Comment (marcoscaceres) [inline index.html]: Maybe clearer:

```suggestion
            <dt>
              <dfn>non-standard</dfn>
            </dt>
            <dd>
              Represents a button that exists but doesn't have a standardized name.
            </dd>
```

File: index.html [hunk 38]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+            <dt>
+              <dfn>trackpad</dfn>
+            </dt>
+            <dd>
+              Represent a trackpad input type which beyonds the scope of the standard mapping.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
              Represent a trackpad input type.
```

File: index.html [hunk 39]
```diff
@@ -2031,6 +2051,62 @@ <h2>
...
+	      <p>More enum types to be defined by the separate W3C wiki page.
```
Comment (marcoscaceres) [inline index.html]: On reflection, we should do this in the spec... 

```suggestion
```

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Some suggested changes.

**Comment by xingri** [comment]
> Some suggested changes.

@marcoscaceres thank you so much for the detailed feedbacks.
Will prepare the revised version based on your feedback soon.

**Comment by xingri** [comment]
> Some suggested changes.

@marcoscaceres Thanks for the feedback. Applied most of the change requests. 

Regarding the extended mapping, based on the working group discussion on last March with @nondebug while you were away, 
we discussed that we will not have the additional mapping type of "extended" for those gamepads that have a mapping that is "standard" plus extended bits.  So, we'd keep "standard" but just have more controls.

Let's discuss a little bit during the call. Also I will fix the syntax error soon.

**Comment by xingri** [comment]
@marcoscaceres, @nondebug  Please review the updated sentences.
Also linked the webkit & chromium issues.

File: index.html [hunk 40]
```diff
@@ -820,8 +820,13 @@ <h3>
           <li>[=list/For each=] |rawInputIndex:long| of [=the range=] from 0 to
           |inputCount| − 1:
             <ol>
-              <li>If the the gamepad button at index |rawInputIndex|
-              [=represents a Standard Gamepad button=]:
+              <li>Let |button| be the button at index |rawInputIndex|.
+              </li>
+              <li>Let |type| be the result of determining if the |button| [=represents
+              a Standard Gamepad button=] or represents [=Additional gamepad buttons=]
```
Comment (nondebug) [inline index.html]: Let's use singular in both cases and add a period at the end of the sentence:

```
a Standard Gamepad button=] or represents an [=additional gamepad button=].
```

I'm not sure if respec will link the definition correctly if we don't use "Additional gamepad buttons"; if not, we can add a data-lt attribute to the <dfn> tag below:

```
<dfn data-lt="additional gamepad button">Additional gamepad buttons</dfn>
```

File: index.html [hunk 41]
```diff
@@ -878,8 +883,15 @@ <h3>
           <li>Initialize |buttons| to be an empty [=list=].
           </li>
           <li>[=list/For each=] |buttonIndex:long| of [=the range=] from 0 to
-          |buttonsSize| − 1, [=list/append=] a [=new=] {{GamepadButton}} to
-          |buttons|.
+          |buttonsSize| − 1:
+            <ol>
+              <li>Determine the |type| of the button from the provided {{GamepadButtonType}}
+              array, defaulting to null if not specified.
```
Comment (nondebug) [inline index.html]:
```
`null`
```

File: index.html [hunk 42]
```diff
@@ -878,8 +883,15 @@ <h3>
...
+              </li>
+              <li>[=list/Append=] a [=new=] {{GamepadButton}} to |buttons|
```
Comment (nondebug) [inline index.html]: This should reference the |type| variable we defined above:

```
<li>Let |button:GamepadButton| be a [=new=] {{GamepadButton}} with its {{GamepadButton/type}} attribute initialized to |type|.
</li>
<li>[=list/Append=] |button| to |buttons|.
</li>
```

File: index.html [hunk 43]
```diff
@@ -878,8 +883,15 @@ <h3>
           <li>Initialize |buttons| to be an empty [=list=].
           </li>
           <li>[=list/For each=] |buttonIndex:long| of [=the range=] from 0 to
-          |buttonsSize| − 1, [=list/append=] a [=new=] {{GamepadButton}} to
-          |buttons|.
+          |buttonsSize| − 1:
+            <ol>
+              <li>Determine the |type| of the button from the provided {{GamepadButtonType}}
```
Comment (nondebug) [inline index.html]: > Determine the |type| of the button from the provided {{GamepadButtonType}} array

This should be phrased "Let |type| be..." so it's clear we're defining a variable. Also, it's unclear where the GamepadButtonType array came from or how it's being used. If it's provided then it should be passed into this algorithm and given a name, for instance: "To initialize buttons for a gamepad with an array |types| of GamepadButtonType values"

Comment (xingri) [reply index.html]: @nondebug  thanks for the review and detailed guide. Updated this PR to reflect the feedbacks.

**Comment by xingri** [comment]
@marcoscaceres could you please review the changes and let me know if you need anything further.

File: index.html [hunk 44]
```diff
@@ -820,8 +820,13 @@ <h3>
           <li>[=list/For each=] |rawInputIndex:long| of [=the range=] from 0 to
           |inputCount| − 1:
             <ol>
-              <li>If the the gamepad button at index |rawInputIndex|
-              [=represents a Standard Gamepad button=]:
+              <li>Let |button| be the button at index |rawInputIndex|.
```
Comment (nondebug) [inline index.html]: Assigning the button to a variable |button| seems awkward. |button| doesn't have a defined type, it's intended to represent a specific physical button under some device-defined button ordering. Let's reword this to avoid assigning it to anything.

```
<li>Let |type| be the result of determining if the button at index |rawInputIndex|
[=represents a Standard Gamepad button=] or...
</li>
```

File: index.html [hunk 45]
```diff
@@ -878,8 +883,14 @@ <h3>
           <li>Initialize |buttons| to be an empty [=list=].
           </li>
           <li>[=list/For each=] |buttonIndex:long| of [=the range=] from 0 to
-          |buttonsSize| − 1, [=list/append=] a [=new=] {{GamepadButton}} to
-          |buttons|.
+          |buttonsSize| − 1:
+            <ol>
+              <li>Let |button:GamepadButton| be a [=new=] {{GamepadButton}} with its
+              {{GamepadButton/type}} attribute initialized to |type|.
```
Comment (nondebug) [inline index.html]: We're initializing button.type but the remaining GamepadButton attributes are left uninitialized. Let's specify that button.value is initialized to zero and button.pressed and button.touched are initialized to false.

File: index.html (hunk 44 above)
Comment (xingri) [reply index.html]: @nondebug  Thanks Matt for the review. Just updated the PR.

File: index.html (hunk 45 above)
Comment (xingri) [reply index.html]: @nondebug  Thanks Matt for the review. Just updated the PR.

File: index.html [hunk 46]
```diff
@@ -878,8 +881,17 @@ <h3>
...
           <li>[=list/For each=] |buttonIndex:long| of [=the range=] from 0 to
-          |buttonsSize| − 1, [=list/append=] a [=new=] {{GamepadButton}} to
-          |buttons|.
+          |buttonsSize| − 1:
+            <ol>
+              <li>Let |button:GamepadButton| be a [=new=] {{GamepadButton}} with its
+              {{GamepadButton/type}} attribute initialized to |type|,
+              button.value initialized to 0,
+              button.pressed initialized to false, and
+              button.touched initialized to false.
```
Comment (nondebug) [inline index.html]: Let's use the same button.attr syntax for all four attributes and reorder them to match the order in the GamepadButton interface definition:

```
              <li>Let |button:GamepadButton| be a [=new=] {{GamepadButton}} with
              |button|.{{GamepadButton/pressed}} initialized to `false`,
              |button|.{{GamepadButton/touched}} initialized to `false`,
              |button|.{{GamepadButton/type}} initialized to |type|, and
              |button|.{{GamepadButton/value}} initialized to 0.
              </li>
```

Comment (xingri) [reply index.html]: @nondebug  Thanks Matt. Just updated.

**Comment by xingri** [comment]
@marcoscaceres , @nondebug  As we discussed today, I have removed the table in the section 12.1.
Please review this PR and let me know your feedback.

File: index.html [hunk 47]
```diff
@@ -2035,6 +2055,60 @@ <h2>
...
+	    {{GamepadButtonType}} enumeration.
+          </p>
+          <p>
+            This enum defines the set of possible button types.
+          </p>
+          <pre class="idl">
+            enum GamepadButtonType {
+              "standard",
+              "non-standard",
+              "trackpad",
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
              "non-standard",
              "standard",
              "trackpad",
```

File: index.html [hunk 48]
```diff
@@ -2035,6 +2055,60 @@ <h2>
           Visual representation of a [=Standard Gamepad=] layout.
         </figcaption>
       </figure>
+      <section>
+        <h3>
+          <dfn data-lt="additional gamepad button">Additional gamepad buttons</dfn>
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          <dfn>Additional gamepad buttons</dfn>
```
